### ISSUE TITLE
M1 Macbook says it's running but can't access WP Homepage

### ISSUE THREAD

**Issue by azatecas** [10up/wp-local-docker-v2#188]
UPDATE: This issue has been resolved through continues Docker updates and Great suggestions from other devs. 

Hello, Im not sure if I should report this as a bug or not because I'm aware of the differences in architecture for the new Apple Silicon and that the available docker installation it's currently a Preview and not official. However I decided to do so because I really enjoy 10updocker and hope other developers may find this useful. 

**Describe the bug**

I have manage to successfully get all the containers in the 10updocker configuration running on the M1 macbook without modifying the docker-compose files.  The installation logs state that the site has been successfully created and it adds the local domain to /etc/hosts file. But when try goin to the domain.test. the browser is unable to find the page. all of the containers are running, including phpmyadmin, nginx:latest, 10up/wp-php-fpm-dev, memcached, mysql, andyshinn/dnsmasq, mailcatcher. Im able to access phpmyadmin by going to 127.0.0.1:8092 and i can create/update databases. however for some reason the WP page isn't available. i tried domain.test, 127.0.0.1 and nothing.  
**Steps to Reproduce**

~~To reproduce these steps its a little involved. the Stock Big Sur/M1 configuration utilizes Port 53 for mDNSResponder for Bonjour. So if you try creating a 10updocker environment youll run into errors that port 53 is already in use. so you need to disable mDNSResponder by doing the following:~~

~~1. Disable System Integrity Protection (SIP) by starting your Mac in recovery mode, accessing the terminal and running `csutil disable`~~
~~2. boot into your profile, start a terminal and type `sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.mDNSResponder.plist` this will disable mDNSResponder and kill the current process  BUT it will break access to your terminal's internet connection. So you must re-enable it with `sudo launchctl load -w /System/Library/LaunchDaemons/com.apple.mDNSResponder.plist`~~
~~3. Now that the mDNSResponder process is not running, make sure no other containers or brew services are running that use the same ports. then create a 10updocker environment with `10updocker create foo.test` and follow intallation steps.~~
1. With the latest version of Docker for M1 (March Release Candidate),  create a 10updocker environment with `10updocker create foo.test` and follow intallation steps
2. The installation will conclude and say its successful, giving you the domain to visit. running `docker ps` will display the active running containers.
**Expected behavior**

when visiting the domain.test, the default WP homepage should appear.
**Screenshots**

![Screen Shot 2021-03-30 at 12 17 35 PM](https://user-images.githubusercontent.com/29685279/113022148-4a032d00-9152-11eb-8f97-a5e217e2d75e.png)

<img width="1334" alt="Screen Shot 2021-03-30 at 12 17 02 PM" src="https://user-images.githubusercontent.com/29685279/113022027-2cce5e80-9152-11eb-9535-4cc0b684b483.png">
<img width="1062" alt="Screen Shot 2021-03-30 at 11 38 17 AM" src="https://user-images.githubusercontent.com/29685279/113022068-38ba2080-9152-11eb-83a9-f14e96659e73.png">

**Environment information**
 - Device: M1 Macbook Air
 - OS: Big Sur 11.2.3
 - Browser and version: Brave, Firefox, Safari
 - WordPress version: default install

**Additional context**
~~I dont recomment you disable System Integrity Protection or mDNSResponder for long periods of time. I just tested this  to get 10updocker running on the default compose file since i kept getting errors when trying to create a wp environment. 
reenable System Integrity Protection (SIP) entering recovery mode, accessing the terminal and running `csutil enable` Im not sure why mDNSResponder needs to be taking up port 53 but i find it annoying.~~
  As previosly stated, Not sure if i should report this or not. but the fact that 10updocker initializes all the containers with the default configuration and not be able to access the WP homepage is weird. Any comments or suggestions are welcomed.

**UPDATE:**
I upgrade Docker to the latest Release for M1 macs and no longer have a port 53 error, all the containers start up However the issue of no being able to access the WP home page remains.

**Comment by nicoladj77** [comment]
@azatecas there is a new RC for Dockers which fixes the issue with port 53, have you tested with that?I am able to get all the coontainer up and running switching to mariadb for the general containers but I get an error 
`: ‘/’ is not writable by current user.`
when downloading windows

**Comment by azatecas** [comment]
> @azatecas there is a[ new RC](https://docs.docker.com/docker-for-mac/apple-m1/) for Dockers which fixes the issue with port 53, have you tested with that?I am able to get all the coontainer up and running switching to mariadb for the general containers but I get an error
> `: ‘/’ is not writable by current user.`
> when downloading windows

I downloaded the March Release of Docker for the M1 and  no longer have a port 53 error. However the same issue applies that the containers are all running but im unable to access the Wordpress Home page through the local domain, or local host address. Im able to use the default mysql image by adding `platform: linux/x86_64` to where the mysql image is being declared in the docker-compose file in `/.wplocaldocker/global/docker-compose.yml`

**Comment by nicoladj77** [comment]
ok so now I'm in your situation, what I noticed from the dashboard is that the global_gateway_1 ( the nginx proxy ) is not running.
The error is 
WARNING: /etc/nginx/dhparam/dhparam.pem was not found. A pre-generated dhparam.pem will be used for now while a new one

is being generated in the background. Once the new dhparam.pem is in place, nginx will be reloaded.

runtime: failed to create new OS thread (have 2 already; errno=22)

fatal error: newosproc


runtime stack:

runtime.throw(0x884500, 0x9)

/usr/local/go1.6/src/runtime/panic.go:530 +0x90

runtime.newosproc(0xc820026000, 0xc820035fc0)

/usr/local/go1.6/src/runtime/os1_linux.go:149 +0x18c

runtime.newm(0x932358, 0x0)

/usr/local/go1.6/src/runtime/proc.go:1513 +0x135

runtime.main.func1()

/usr/local/go1.6/src/runtime/proc.go:125 +0x2c

runtime.systemstack(0xa8e800)

/usr/local/go1.6/src/runtime/asm_amd64.s:291 +0x79

runtime.mstart()

/usr/local/go1.6/src/runtime/proc.go:1048


goroutine 1 [running]:

runtime.systemstack_switch()

/usr/local/go1.6/src/runtime/asm_amd64.s:245 fp=0xc820020770 sp=0xc820020768

runtime.main()

/usr/local/go1.6/src/runtime/proc.go:126 +0x62 fp=0xc8200207c0 sp=0xc820020770

runtime.goexit()

/usr/local/go1.6/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc8200207c8 sp=0xc8200207c0

**Comment by tigunov-it** [comment]
Hello!
How did you run mysql container on the M1 processor?  
Error : Pulling mysql (mysql:5)...
no matching manifest for linux/arm64/v8 in the manifest list entries.
Is it possible to change the wp-local-docker-v2 config to work with mariadb?

**Comment by jeremiahbratton** [comment]
Bumping. Wondering if anyone has seen this resolved for anyhow, or if perhaps it was resolved in another thread I am not seeing as I am having the same issue. 

I have a M1 Mac mini fresh out of the box. Got docker and 10updocker setup. Everything is on current versions. I am able to get through the setup process but cannot access the local site after that through the urls I provide (ex: sandbox.test). However I am able to access the tools such as phpMyAdmin. `docker ps` shows that things are in the right place and host file entries are getting written as expected.

I did have to add platform: `linux/x86_64` to get the install running

and these are the versions
Big Sur 11.3.1
Docker 3.3.2 (63878)
10updocker 2.8.1

**Comment by nicoladj77** [comment]
@jeremiahbratton so the workaround for now seems to be (tested by others):

In `~/.wplocaldocker/global/docker-compose.yml:`

1. change mysql image from mysql to mariadb
2. change dns gateway image from 10up/nginx-proxy to nginxproxy/nginx-prox

**Comment by jeremiahbratton** [comment]
Updated the file and now when I use `start` or `create` getting hung on _Ensuring global services are running..._

I might have also just taken the 2 changes too directly. So it is very likely that I missed a step somewhere or put something in the wrong place. Here is the dns and mysql portion of the file if you don't mind double checking me. 

@nicoladj77 Also, I don't usually find myself posting in issues for help (as the questions are usually not new and have been solved) If you have a recommendation for a place that I check in on to keep up with this issue I will take it. And, further in advance thanks for the time.

```
  dns:
    image: andyshinn/dnsmasq
    cap_add:
      - NET_ADMIN
    command: -A /test/10.0.0.3 --log-facility=-
    ports:
        - 127.0.0.1:53:53/udp
        - 127.0.0.1:53:53
    networks:
      wplocaldocker:
        ipv4_address: 10.0.0.2
  gateway:
    image: nginxproxy/nginx-prox
    ports:
      - 127.0.0.1:80:80
      - 127.0.0.1:443:443
    volumes:
      - ./ssl-certs:/etc/nginx/certs:ro
      - /var/run/docker.sock:/tmp/docker.sock:ro
    networks:
      wplocaldocker:
        ipv4_address: 10.0.0.3
  mysql:
    image: mariadb
    volumes:
      - ./config/mysql/config.cnf:/etc/mysql/mysql.conf.d/wp-local-docker.cnf:ro
      - mysqlData:/var/lib/mysql
    ports:
      - 127.0.0.1:3306:3306
    environment:
      MYSQL_ROOT_PASSWORD: password
      MYSQL_USER: wordpress
      MYSQL_PASSWORD: password
    mem_limit: 1024M
    mem_reservation: 1024M
    networks:
      - wplocaldocker
```

**Comment by jeremiahbratton** [comment]
An update to share here:
nginx-proxy was updated today https://github.com/10up/wp-local-docker-images/issues/21#issuecomment-833068354

and there is a call for field testing in the issue

**Comment by jeremiahbratton** [comment]
I was able to get 10updocker running on my M1 Mac Mini today after the update to nginx-proxy went up. I want to recount the steps I took to get it running. This is not fully official but, if like me, you need your tools back... any port in a storm.

1. Run `10updocker stop all` and make sure all docker containers that are part of wp-local-docker global servers are destroyed (check `docker ps -a`).
2. Edit the '~/.wplocaldocker/global/docker-compose.yml' file and update the image entry for the gateway from '10up/nginx-proxy' to '10up/nginx-proxy:multiarch'
3. Change `image: mysql:5` to `image: mariadb`
4. If you had previous put `platform: linux/x86_64` under your mysql settings remove it.
5. `start` or `create` an environment

**Notes:** 

- I had a few conflicts getting this running at first. I had the mysql and nginx-proxy containers misfire. If you get port conflicts or docker is getting crashy I recommend opening up the dashboard for docker desktop and deleting the `global_mysql_1` and `global_gateway_1` container manually. After they are clear run `10updocker stop all` and then attempt another `start` or `create`
- The mysql to mariaDB image switch is not the official path forward. This is a recommendation that I got from others. I was able to get it up and working. Its an in the moment patch, results may vary. Just remember that you made the change if things stop working in the near future.

**Comment by azatecas** [comment]
@nicoladj77 @jeremiahbratton You guys are awesome, thanks to you guys i have been able to get 10updocker working locally on the M1 macbook air. Thanks for the steps @jeremiahbratton they were super helpful and very descriptive. Now i can stop using MAMP

**Comment by tigunov-it** [comment]
Hello! How did you solve the problem of port 53 on a mac os? 
"ERROR: for global_dns_1  Cannot start service dns: Ports are not available: port is already allocated"

**Comment by azatecas** [comment]
> Hello! How did you solve the problem of port 53 on a mac os? 
> "ERROR: for global_dns_1  Cannot start service dns: Ports are not available: port is already allocated"

The latest docker release candidate fixes the port 53 issue. As of right now the latest release is from April 15th. you should do a clean install and see if it works.

**Comment by dustinrue** [comment]
The work around for this, and honestly it'll most likely become _the fix_ is to remove the port entries for the dnsmasq container in `~/.wplocaldocker/global/docker-compose.yml` and restart everything

**Comment by gautam-jha** [comment]
> I was able to get 10updocker running on my M1 Mac Mini today after the update to nginx-proxy went up. I want to recount the steps I took to get it running. This is not fully official but, if like me, you need your tools back... any port in a storm.
> 
> 1. Run `10updocker stop all` and make sure all docker containers that are part of wp-local-docker global servers are destroyed (check `docker ps -a`).
> 2. Edit the `~/.wplocaldocker/global/docker-compose.yml` file and update the image entry for the gateway from `10up/nginx-proxy` to `10up/nginx-proxy:multiarch`
> 3. Change `image: mysql:5` to `image: mariadb`
> 4. If you had previous put `platform: linux/x86_64` under your mysql settings remove it.
> 5. `start` or `create` an environment
> 
> **Notes:**
> 
> * I had a few conflicts getting this running at first. I had the mysql and nginx-proxy containers misfire. If you get port conflicts or docker is getting crashy I recommend opening up the dashboard for docker desktop and deleting the `global_mysql_1` and `global_gateway_1` container manually. After they are clear run `10updocker stop all` and then attempt another `start` or `create`
> * The mysql to mariaDB image switch is not the official path forward. This is a recommendation that I got from others. I was able to get it up and working. Its an in the moment patch, results may vary. Just remember that you made the change if things stop working in the near future.

This solution for Arch Linux works perfectly.

**Comment by azatecas** [comment]
I wanted to follow up this thread since it's been super helpful. I have been able to increase 10updocker performance on the M1 Apple Silicon Chip by replacing some of the default images that were running in x86 emulation to arm builds and disabling some containers that i did not need.

@jeremiahbratton recommendations listed above are the first steps. 
Swap the **mysql** image for **mariadb** and replace **10up/nginx-proxy** with **10up/nginx-proxy:multiarch**

you can also replace **andyshinn/dnsmasq** image for **gabrielaraujof/dnsmasq-armhf**, it appears to be an arm build of andyshinn's dns masq image.

you can also replace **10up/wp-php-fpm-dev:7.4** with **10up/wp-php-fpm-dev:7.4-ubuntu** the ubuntu build appears to run the arm binary, while the default goes through rossetta emulation.

since i use Sequel Pro for managing databases i dont need the phpmyadmin image running so i commented it out on the docker-compose.yml file and freed some resources from the cpu. I did the same for mailcatcher since the current sites i've been working dont need this functionality locally. 

After doing these changes the differences is night and day. you may not be able to tell the performance difference if you are running a basic install of Wordpress but if you have a large site running locally you will notice. I cant say which of the edits above contributes the most to performance. maybe just disabling unwanted containers was enough. but not having a single container run through rosetta definitively uses less cpu resources.

Keep up the great work 10up Devs you guys rock!

here is my yml file that lives in $HOME/.wplocaldocker/global/docker-compose.yml
`version: "2.4"
services:
  # Run a local dnsmasq server for the containers that will resolve DNS queries and
  # route any *.test domains to the gateway container
  # dns:
  #   image: andyshinn/dnsmasq
  #   cap_add:
  #     - NET_ADMIN
  #   command: -A /test/10.0.0.3 --log-facility=-
  #   ports:
  #       - 127.0.0.1:53:53/udp
  #       - 127.0.0.1:53:53
  #   networks:
  #     wplocaldocker:
  #       ipv4_address: 10.0.0.2
  
  dns:
    image: gabrielaraujof/dnsmasq-armhf
    cap_add:
      - NET_ADMIN
    command: -A /test/10.0.0.3 --log-facility=-
    ports:
        - 127.0.0.1:53:53/udp
        - 127.0.0.1:53:53
    networks:
      wplocaldocker:
        ipv4_address: 10.0.0.2
  gateway:
    image: 10up/nginx-proxy:multiarch
    ports:
      - 127.0.0.1:80:80
      - 127.0.0.1:443:443
    volumes:
      - ./ssl-certs:/etc/nginx/certs:ro
      - /var/run/docker.sock:/tmp/docker.sock:ro
    networks:
      wplocaldocker:
        ipv4_address: 10.0.0.3
  mysql:
    image: mariadb
    volumes:
      - ./config/mysql/config.cnf:/etc/mysql/mysql.conf.d/wp-local-docker.cnf:ro
      - mysqlData:/var/lib/mysql
    ports:
      - 127.0.0.1:3306:3306
    environment:
      MYSQL_ROOT_PASSWORD: password
      MYSQL_USER: wordpress
      MYSQL_PASSWORD: password
    mem_limit: 1024M
    mem_reservation: 1024M
    networks:
      - wplocaldocker
  # mailcatcher:
  #   image: schickling/mailcatcher
  #   ports:
  #     - 127.0.0.1:1025:1025
  #     - 127.0.0.1:1080:1080
  #   environment:
  #     MAILCATCHER_PORT: 1025
  #   networks:
  #     - wplocaldocker
  # phpmyadmin:
  #   image: phpmyadmin/phpmyadmin
  #   ports:
  #     - 127.0.0.1:8092:80
  #   depends_on:
  #     - mysql
  #   environment:
  #     PMA_HOST: mysql
  #   networks:
  #     - wplocaldocker
networks:
  wplocaldocker:
    external:
      name: wplocaldocker
volumes:
  mysqlData:
  elasticsearchData:
`

and here is the docker-compose file that lives in $HOME/wp-local-docker/{your-project-name}/docker-compose.yml

`version: '2.2'
services:
  memcached:
    image: 'memcached:latest'
  nginx:
    image: 'nginx:latest'
    expose:
      - '80'
      - '443'
    depends_on:
      - phpfpm
    networks:
      - default
      - wplocaldocker
    volumes:
      - './wordpress:/var/www/html:cached'
      - './config/nginx/server.conf:/etc/nginx/conf.d/common/_server.conf:cached'
      - './config/nginx/default.conf:/etc/nginx/conf.d/default.conf:cached'
    environment:
      CERT_NAME: test-test
      HTTPS_METHOD: noredirect
      VIRTUAL_HOST: 'test.test,*.test.test'
  phpfpm:
    image: '10up/wp-php-fpm-dev:7.4'
    depends_on:
      - memcached
    networks:
      - default
      - wplocaldocker
    dns:
      - 10.0.0.2
    volumes:
      - './wordpress:/var/www/html:cached'
      - './config/php-fpm/docker-php-ext-xdebug.ini:/etc/php.d/docker-php-ext-xdebug.ini:cached'
      - 'wplocaldockerCache:/var/www/.wp-cli/cache:cached'
      - '~/.ssh:/home/www-data/.ssh:cached'
      - './config/php-fpm/wp-cli.local.yml:/var/www/.wp-cli/config.yml:cached'
    cap_add:
      - SYS_PTRACE
    environment:
      ENABLE_XDEBUG: 'true'
      PHP_IDE_CONFIG: serverName=test-test
networks:
  wplocaldocker:
    external:
      name: wplocaldocker
volumes:
  wplocaldockerCache:
    external:
      name: wplocaldockerCache
`

**Comment by azatecas** [comment]
WP Local Docker 3.0 has been release, it includes arm optimized images and support for PHP 8.0. The modification above is no longer needed.
